### DISCUSSION TITLE
Make events consumable

### DISCUSSION THREAD

**Discussion by JustRed23** [discord-jda/JDA#2387] [Feedback & Suggestions]
I would like to add a feature that adds consumable events. We have an event listener and we can process all events there but why not in the Button itself?

Ex. `Button.primary("buttonid:dostuff", "Do stuff").onEvent(buttonInteractionEvent -> buttonInteractionEvent.getChannel().sendMessage("Hello world!").queue())`

You could still listen for the events but they could get handled like this as well

**Comment by MinnDevelopment** [comment]
> but why not in the Button itself?

Because that would result in a massive amount of memory allocations. You are creating a new listener every single time you create a button.

Also, the event system in JDA is pluggable, and we cannot provide a consistent way to do this on buttons with the different event systems. Additionally, buttons are created using **static** factory methods that do not have any access to the gateway context.

**Comment by JustRed23** [comment]
Interesting, thanks
